nav2_costmap_2d/include/nav2_costmap_2d/image_processing.hpp

```diff
@@ -0,0 +1,1047 @@
+ * Dereferences a pointer to a existing pixel. If the pixel is out of bounds, it returns a ref to 0.
+ * @tparam T image pixel type
+ * @sa DoNothing
+ */
+template<class T>
+class ReplaceToZero
+{
+public:
+  /**
+   * @brief Create an object that will replace pointers outside the specified range
+   * @param up_row_start pointer to the first pixel of up row. Can be nullptr.
+   * @param down_row_start pointer to the first pixel of down row
+   * @param columns number of pixels in both rows
+   */
+  ReplaceToZero(const T * up_row_start, const T * down_row_start, size_t columns)
+  : up_row_start_{up_row_start}, up_row_end_{up_row_start + columns},
+    down_row_start_{down_row_start}, down_row_end_{down_row_start + columns} {}
+
+  /**
+   * @brief Return ref to pixel or to zero value if up_row_start_ is nullptr or the pointer is out of bounds
+   * @param v pointer to pixel
+   */
+  T & up(T * v)
+  {
+    if (up_row_start_ == nullptr) {
+      return zero_;
+    }
+    return replaceOutOfBounds(v, up_row_start_, up_row_end_);
+  }
+
+  /**
+   * @brief Return ref to pixel or to zero value if the pointer is out of bounds
+   * @param v pointer to pixel
+   */
+  T & down(T * v)
+  {
+    return replaceOutOfBounds(v, down_row_start_, down_row_end_);
+  }
+
+private:
+  /**
+   * @brief Replaces an out-of-bounds pointer with a pointer to 0
+   * @return a dereferenced pointer or a reference to 0 if the pointer is out of range
+   */
+  T & replaceOutOfBounds(T * v, const T * begin, const T * end)
+  {
+    if (v < begin || v >= end) {
+      return zero_;
+    }
+    return *v;
+  }
+
+  const T * up_row_start_;
+  const T * up_row_end_;
+  const T * down_row_start_;
+  const T * down_row_end_;
+  T zero_{};
+};
+
+}  // namespace out_of_bounds_policy
+
+/**
+ * @brief Forward scan mask sliding window
+ * Provides an interface for access to neighborhood of the current pixel
+ * (includes three neighbors of the top row, the pixel to the left and the current one).
+ * In the illustration below, the current pixel is e.
+ * |a|b|c|
+ * |d|e| |
+ * | | | |
+ * @tparam T image pixel type
+ * @tparam Border optional check of access to pixels outside the image boundary (DoNothing or ReplaceToZero)
+ */
+template<class T, template<class> class Border>
+class Window
+{
+public:
+  /**
+   * Construct mask window
+   * @param up_row pointer to the pixel above the current one (a)
+   * @param down_row pointer to the current pixel (e)
+   * @param border boundary case object
+   */
+  inline Window(T * up_row, T * down_row, Border<T> border = {})
+  : up_row_{up_row}, down_row_{down_row}, border_{border} {}
+
+  inline T & a() {return border_.up(up_row_ - 1);}
+  inline T & b() {return border_.up(up_row_);}
+  inline T & c() {return border_.up(up_row_ + 1);}
+  inline T & d() {return border_.down(down_row_ - 1);}
+  inline T & e() {return border_.down(down_row_);}
+  inline const T * anchor() const {return down_row_;}
```

Comment:
Yes, it's the same. And the check in line 308 seems redundant. I'll check and fix.

Reply:
[Fixed](https://github.com/ros-planning/navigation2/pull/2567/commits/e4a9d49f0f820314866afa470a1ccba66e58938f). As for some redundancy: I left both functions, since it is desirable to syntactically distinguish between access to the window element `e` (with the same symbolic name as in the reference article) and the window anchor, which are used in different contexts. In addition, they have different const requirements.

Reply:
Okay, accepted. BTW, `inline T & b() {return border_.up(up_row_);}` using this logic is also could be updated to `inline T & b() {return *up_row_;}`

Reply:
No, checking when accessing any element of the top row is necessary. At [the stage](https://github.com/ros-planning/navigation2/blob/b728c4c7492f600f490aabf29bc213c1a11057e1/nav2_costmap_2d/include/nav2_costmap_2d/image_processing.hpp#L817) of processing the top image row with index 0, the center of the window corresponds to the zero image row. The top line of the window corresponds to the non-existent line (with index -1 ). Therefore, access to elements a, b, c in the general case should be checked.

Reply:
Got it: `img.e()` (or `img.anchor()`) is the only one instance that could be safely addressed directly w/o any out-of-border checks.